File: backend/calculations.py
"""
Turns (items + claims + tax/tip) into a final amount owed per participant.

Rules:
  1. An item claimed by N people is split evenly N ways.
  2. An item claimed by nobody is split evenly across ALL participants
     (fallback so the receipt always fully reconciles — no penny left behind).
  3. Tax and tip are allocated proportionally to each person's share of
     the subtotal (fair: heavy orderers pay more tax, not equal split).
  4. Rounding remainder is absorbed by the payer so the payer's total
     always matches the receipt total exactly.
"""
from collections import defaultdict


def calculate_splits(
    items: list[dict],
    claims: list[dict],
    tax: float,
    tip: float,
    participants: list[str],
    payer_participant_id: str,
) -> dict[str, dict]:
    """
    Args:
        items:                 list of {id, name, price, quantity}
        claims:                list of {item_id, participant_id}
        tax:                   total tax from receipt
        tip:                   total tip from receipt
        participants:          list of participant ids in the session
        payer_participant_id:  participant who paid the restaurant

    Returns:
        {participant_id: {subtotal, tax_share, tip_share, total}}
    """
    if not participants:
        return {}

    # Build a map: item_id -> [participant_ids who claimed it]
    claims_by_item: dict[str, list[str]] = defaultdict(list)
    for c in claims:
        claims_by_item[c["item_id"]].append(c["participant_id"])

    subtotal_by_participant: dict[str, float] = defaultdict(float)
    total_subtotal = 0.0

    for item in items:
        price = float(item["price"])
        total_subtotal += price
        claimants = claims_by_item.get(item["id"], [])

        if not claimants:
            # Nobody claimed it — split among everyone (unclaimed fallback)
            share = price / len(participants)
            for p in participants:
                subtotal_by_participant[p] += share
        else:
            share = price / len(claimants)
            for p in claimants:
                subtotal_by_participant[p] += share

    result: dict[str, dict] = {}

    for p in participants:
        p_subtotal = subtotal_by_participant.get(p, 0.0)
        proportion = (p_subtotal / total_subtotal) if total_subtotal > 0 else (1 / len(participants))
        p_tax = tax * proportion
        p_tip = tip * proportion
        p_total = p_subtotal + p_tax + p_tip

        result[p] = {
            "subtotal": round(p_subtotal, 2),
            "tax_share": round(p_tax, 2),
            "tip_share": round(p_tip, 2),
            "total": round(p_total, 2),
        }

    # Fix rounding: diff between receipt total and sum of ROUNDED totals
    # (rounding each total to 2dp can lose/gain a cent — payer absorbs it)
    true_total = round(total_subtotal + tax + tip, 2)
    rounded_sum = round(sum(r["total"] for r in result.values()), 2)
    if payer_participant_id in result:
        diff = round(true_total - rounded_sum, 2)
        result[payer_participant_id]["total"] = round(
            result[payer_participant_id]["total"] + diff, 2
        )

    return result

File: proj/src/lib/calculations.ts
/**
 * Port of backend/calculations.py -> TypeScript
 */

type Item = { id: string; name?: string; price: number; quantity?: number };
type Claim = { item_id: string; participant_id: string };

export function calculateSplits(
  items: Item[],
  claims: Claim[],
  tax: number,
  tip: number,
  participants: string[],
  payerParticipantId: string
): Record<string, { subtotal: number; tax_share: number; tip_share: number; total: number }> {
  if (!participants || participants.length === 0) return {};

  const claimsByItem = new Map<string, string[]>();
  for (const c of claims || []) {
    claimsByItem.set(c.item_id, (claimsByItem.get(c.item_id) || []).concat(c.participant_id));
  }

  const subtotalByParticipant = new Map<string, number>();
  let totalSubtotal = 0;

  for (const item of items || []) {
    const price = Number(item.price) || 0;
    const quantity = item.quantity ?? 1;
    const lineTotal = price * quantity;
    totalSubtotal += lineTotal;
    const claimants = claimsByItem.get(item.id) || [];
    if (claimants.length === 0) {
      const share = lineTotal / participants.length;
      for (const p of participants) {
        subtotalByParticipant.set(p, (subtotalByParticipant.get(p) || 0) + share);
      }
    } else {
      const share = lineTotal / claimants.length;
      for (const p of claimants) {
        subtotalByParticipant.set(p, (subtotalByParticipant.get(p) || 0) + share);
      }
    }
  }

  const result: Record<string, { subtotal: number; tax_share: number; tip_share: number; total: number }> = {};

  for (const p of participants) {
    const pSubtotal = subtotalByParticipant.get(p) || 0;
    const proportion = totalSubtotal > 0 ? pSubtotal / totalSubtotal : 1 / participants.length;
    const pTax = tax * proportion;
    const pTip = tip * proportion;
    const pTotal = pSubtotal + pTax + pTip;
    result[p] = {
      subtotal: Math.round(pSubtotal * 100) / 100,
      tax_share: Math.round(pTax * 100) / 100,
      tip_share: Math.round(pTip * 100) / 100,
      total: Math.round(pTotal * 100) / 100,
    };
  }

  const trueTotal = Math.round((totalSubtotal + tax + tip) * 100) / 100;
  const roundedSum = Math.round(Object.values(result).reduce((s, r) => s + r.total, 0) * 100) / 100;
  if (payerParticipantId in result) {
    const diff = Math.round((trueTotal - roundedSum) * 100) / 100;
    result[payerParticipantId].total = Math.round((result[payerParticipantId].total + diff) * 100) / 100;
  }

  return result;
}
